types.bal

```diff
@@ -408,46 +435,73 @@ public isolated function mapToTimelineStepResponse(
         int experimentId,
+        int timelineStepSequence,
         database:TimelineSubstepWithParams substep,
         database:ExperimentDataReference[] inputData = []
-) returns TimelineSubstepResponse {
+) returns TimelineSubstepResponseWithParams {
     var inputDataLinks = from var dataRef in inputData
         select string `/experiments/${experimentId}/data/${dataRef.name}?version=${dataRef.'version}`;
     return {
-        '\@self: string `/experiments/${experimentId}/timeline/${substep.substepNr}`,
-        stepId: substep.stepId,
+        '\@self: string `/experiments/${experimentId}/timeline/${timelineStepSequence}/substeps/${timelineStepSequence}`,
+        stepId: timelineStepSequence,
+        sequence: timelineStepSequence,
         substepId: substep.substepId,
         substepNr: substep.substepNr,
         href: substep.href,
         hrefUi: substep.hrefUi,
         cleared: substep.cleared == 1 ? true : false,
-        parameters: string `/experiments/${experimentId}/timeline/${substep.stepId}/substeps/${substep.substepNr}/parameters`,
+        parameters: string `/experiments/${experimentId}/timeline/${timelineStepSequence}/substeps/${substep.substepNr}/parameters`,
         parametersContentType: substep?.parametersContentType,
         inputData: inputData,
         inputDataLinks: inputDataLinks
     };
 }
 
+# Helper function to map `TimelineSubstepSQL[]` database records to API records.
+#
+# + experimentId - the id of the experiment the substep is part of
+# + timelineStepSequence - the sequence number of the associated timeline step
+# + substeps - the input substep list
+# + return - the mapped record
+public isolated function mapToTimelineSubstepListResponse(
+        int experimentId,
+        int timelineStepSequence,
+        database:TimelineSubstepSQL[] substeps
+) returns TimelineSubstepResponseWithoutParams[] {
+    return from var substep in substeps
+        select {
+            stepId: timelineStepSequence,
+            sequence: timelineStepSequence,
+            substepId: substep.substepId,
+            substepNr: substep.substepNr,
+            href: substep.href,
+            hrefUi: substep.hrefUi,
+            cleared: substep.cleared == 1 ? true : false,
+            inputData: substep?.inputData
+        };
+}
+
 # Parse a data input URL to extract the URL parameters.
 #
 # + experimentId - the current experiment id
 # + url - the data input URL to parse
 # + return - the parsed parameters
 public isolated function mapFileUrlToDataRef(int experimentId, string url) returns database:ExperimentDataReference|error {
     // TODO refactor once regex supports extrating group matches
-    var regex = string `^(https?:\/\/)?[^\/]*\/experiments\/${experimentId}\/data\/[^\/]+\/download\?version=(latest|[0-9]+)$`;
+    var regex = string ` ^ (https ?: \/\/) ? [ ^ \/] * \/experiments\/ ${experimentId}\/data\/[^\/]+\/download\?version = (latest | [0 - 9] + )$`;
```

Comment:
check if reformatting breaks this regex (e.g. if the spaces are ignored or interpreted...)

Reply:
weird, not sure why that changed...